Implement initial v2 Package Manager

hasheddan opened the pull request:
<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->

Implementation of the new package manager.

TODO:
- [x] Complete unit tests (this PR)
- [x] Clean up commits (this PR)
- [x] Restrict what a `ProviderRevision` and `ConfigurationRevision` can install (this PR -- linting packages)
- [x] Update helm chart with required permissions (this PR -- related to #1706)
- [x] Remove `replace` stanza for `crossplane-runtime` (this PR)
- [x] Add more descriptive conditions (this PR)
- Update design doc (follow-up PR)
- Set extra permission requests in revision status (follow-up PR?)
- Restrict based on crossplane version restriction (follow-up PR)
- Add hosted mode support (follow-up PR -- tracked by https://github.com/crossplane/crossplane/issues/1691)

Fixes https://github.com/crossplane/crossplane/issues/1690

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

Manual testing artifacts will be added below.

[contribution process]: https://git.io/fj2m9

hasheddan commented on apis/pkg/v1alpha1/provider_types.go:
```diff
@@ -0,0 +1,105 @@
+// ProviderSpec specifies details about a request to install a provider to
+// Crossplane.
+type ProviderSpec struct {
+	ControllerOptions `json:",inline"`
```
`ProviderSpec` differs from `ConfigurationSpec` in that it uses this `ControllerOptions` struct, which is also part of the `PackageRevisionSpec`. The `ConfigurationSpec` has the `ImagePullSecrets` and `ImagePullPolicy` parts of the `ControllerOptions`, but not the `ServiceAccountOptions` because the `Configuration` doesn't produce a controller that needs an SA.

We are going to need to allow configuration of the Provider's packaged controller Deployment, but I think it should be completely separate from the configuration of the unpack Job. The first step to get here would be to add `ImagePullSecrets` and `ImagePullPolicy` to `ProviderSpec` outside of the `ControllerOptions` struct. These would be the specific fields for configuring the unpack Job, just like they are in `ConfigurationSpec`.

This leaves the need to determine how to allow a user to pass configuration to the `Deployment`. We have discussed a few approaches in the past, but haven't landed on a clear direction, mostly because we don't have enough data on how folks will want to configure providers. I think we can confidently say the following:

1. Passing configuration to a Provider deployment is a desirable feature.
2. Allowing passing of configuration to a Provider deployment is a potential major security risk to anyone hosting a Crossplane. This likely means the package manager needs to be able to turn this feature on or off, similar to how it does now.
3. The default Provider deployment should be as secure as possible given the typical use case.
4. Configuration should be passed at package "install" time rather than package "build" time. If there is a configuration that is required for every deployment of a specific Provider, the package author should specify that in the documentation, and users should provide that configuration. This does require a bit of overhead, but results in a much simpler model where we don't have to worry about reconciling packaged configuration with install-time configuration. This could always be expanded in the future.
5. Configuring Provider deployments _could_ be deferred to after this initial implementation. The only known use-case right now is with IAM Roles for SA when provider-aws is installed into a Crossplane cluster running on EKS, and we are currently planning on keeping the `ClusterPackageInstall` around for a few releases, so `provider-aws` could still be installed to work in that way. Also, though it would be a degraded experience, a Provider can always be installed manually without the package manager.

With these in mind, I think I would support starting without configurability, then talking to some users and deciding the best way to support it. Alternatively, we could allow a full `DeploymentSpec` in the `Provider` install Spec, then just consider or ignore that based on how the package manager is configured, but I would prefer we bias towards not adding fields until we know how we want them to be used. The cost of adding is much less than the cost of removing.

How does all of this sound? /cc @lukeweber @negz

hasheddan replied:
> While I agree that folks could work around the issue by manually deploying the provider, I suspect it would be a better UX if they didn't have to.

yes, we definitely need to make this a good experience, but it seems that there are some barriers to doing so right now that may take time to work out in a long-term sustainable way. One other thing to mention here as a use-case is that the `imagePullPolicy` and `imagePullSecrets` also flow down to the `Deployment` right now. This allows you to use an image from a private registry for the controller `Deployment`. The uses of this are likely fairly limited right now because this is only applicable for Providers and I am not sure how many proprietary Providers are in existence. However, we could allow this to continue to flow down for now, and just document the behavior change in the future.

hasheddan replied:
Ping @turkenh. We should discuss what you see is minimum requirements for initial implementation here.

negz replied:
I believe after discussion with @turkenh we're good to proceed without exposing any options to configure the provider (for now), right?

hasheddan replied:
Correct!

hasheddan commented on pkg/controller/pkg/revision/deployment.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+func buildProviderDeployment(provider *metav1alpha1.Provider, revision v1alpha1.PackageRevision, namespace string) (*corev1.ServiceAccount, *appsv1.Deployment) {
+	s := &corev1.ServiceAccount{
```
Note: will need to be named in accordance with the RBAC design in #1706

hasheddan commented on pkg/controller/pkg/manager/reconciler.go:
```diff
@@ -0,0 +1,428 @@
+	// package. If we move to parsing the image manifest from registry directly
+	// then we can instead create the package revision and let it manage the
+	// full lifecycle of its job.
+	jobName := strings.Replace(strings.Replace(p.GetSource(), "/", "-", -1), ":", "-", -1)
```
This is a bit gross but gets the job done. In other places we have deferred to naming the "thing" associated with a package revision the hash of its image. However, we can't do that here because we don't know the hash yet. If we adjust the names of things like a Provider deployment or SA, would be good to follow that convention here.

hasheddan replied:
(Obviously only to the extent that we can without using the hash)

hasheddan replied:
@negz yep, but we would also collide at the `_____Revision` level, so likely okay to go ahead and hit that here.

negz replied:
Would that be true even if a config and a package both tried to use the same image? It could be that that’s an obscure enough edge case that we shouldn’t bother with it.

hasheddan replied:
So there are a few cases here (I think no. 4 is what you are getting at):
1. Two providers use the same image name/tag with the same contents (hash). Right now they would collide on the `Job` and the second would be unable to install. If we made the `Job` unique, they would not collide, but would collides when the second tried to create the `ProviderRevision.
2. Two providers use the same image name/tag with different contents (hash). Right now they would collide on the `Job` and the second would be unable to install. If we made the `Job` unique, they would not collide and the `ProviderRevision` would not collide, making the second able to be installed as well (assuming no conflicting objects). However, this could be a somewhat confusing scenario and I am not sure we want to support it.
3. A provider and configuration use the same image name/tag with the same contents (hash). Right now they would collide on `Job`. If we made the `Job` unique, they would not collide until they got to installing objects. One of them would have to have a mismatch with the meta package type in the package, which would cause it to fail to install.
4. A provider and configuration use the same image name/tag with different contents (hash). Right now they would collide on `Job`. If we made `Job` unique, they would not collide, and they would not collide on `___Revision` or when installing objects (assuming the contents fundamentally changed the package type). This would technically be a valid use-case, but I once again am skeptical that we should support it. Using image hashes allow us to accommodate situations in which a package's contents change for the image with the same tag, and a user wants to update to using the new image contents (this could be done by deleting the existing `Job`). However, I don't think this is going to be a common use-case, and I don't think we should further complicate the model by allowing the use of the same image with different contents outside the scope of a single `Provider` or `Configuration`.

negz replied:
If we go with the `providername-abbreviatedhash` naming scheme, will that take care of all of these cases?

hasheddan replied:
Yes 👍

negz commented on pkg/controller/pkg/revision/reconciler.go:
```diff
@@ -0,0 +1,565 @@
+	installNamespace := os.Getenv(envPodNamespace)
+	if installNamespace == "" {
+		return errors.New("cannot initialize controller with empty install namespace")
+	}
```
Would it be possible to take this as a CLI flag rather than an environment variable? I feel that flags are somewhat more discoverable, and kingpin lets us set them from an env var so they're sort of the best of both worlds. They're slightly annoying to plumb down, but there is precedent; https://github.com/crossplane/crossplane/blob/0b4227e/cmd/crossplane/package/manage/manage.go#L95

hasheddan replied:
@negz yep! Works for me. I went with env var for testing purposes, but if we are plumbing it down then that works even better 👍

negz commented on pkg/controller/pkg/revision/reconciler.go:
```diff
@@ -0,0 +1,565 @@
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return errors.Wrap(err, "failed to initialize host clientset with in cluster config")
+	}
```
Do we need to pass a clientset to `NewReconciler`, or could it build one using `mgr.GetConfig()`?

hasheddan replied:
This was preemptive as we are going to need to manipulate building the clientset for host-aware vs regular mode. We _could_ pass information to `NewReconciler` to do that for us, but not sure that would be much of an optimization.

negz replied:
@hasheddan I recall you mentioned you were on board with this approach. Are you planning to address that in this PR?

hasheddan replied:
@negz ah I had updated `SetupProviderRevision` but not `SetupConfigurationRevision` to use `mgr.GetConfig` :+1:

negz commented on pkg/controller/pkg/revision/reconciler.go:
```diff
@@ -0,0 +1,565 @@
+			pr.SetConditions(runtimev1alpha1.Unavailable(), runtimev1alpha1.ReconcileSuccess())
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+		}
+	}
```
I forget whether we discussed during design what would happen to managed resources that are 'orphaned' when a provider revision becomes inactive. I recall that their CRDs and CRs would continue to exist and be owned by an inactive package revision, but I suspect we'd be unable to delete any dangling CRs without a controller running to reconcile them since nothing would remove the finalizers.

hasheddan replied:
So as long as the revision exists, its objects will also exist (the inactive revision is now an Owner instead of Controller). You are correct that in order to clean up CRs after that, you would need to manually remove the finalizer. However, you could also just delete the revision, in which case the CRD would be GC'ed by k8s (no more owner refs), and all CRs for that CRD will also be cleaned up.

I think this is a good pattern as it makes use of kubernetes inherent functionality, rather than adding a bunch of logic on top. The one case where this could be difficult is when you update a package and then want to go and delete some but not all of the CRs for CRDs only owned by the old package revision (this would be a strange case because you would be electing to keep around CRs that were not being reconciled). In this case, I think we should be okay saying that it takes some manual work, and we could even add some utilities to `crank` to help with removing finalizers for things owned by an inactive package revision.

negz replied:
I agree with the approach you propose, but are you sure that deleting a CRD will delete the CRs it defines, even if they have finalizers that never get removed?

hasheddan replied:
You are correct, the CRD will hang around as long as the CR still has a finalizer, so folks would need to go back and manually remove that finalizer (ref: https://github.com/kubernetes/kubernetes/issues/60538#issuecomment-680279481). This is not straightforward for us to address though. If you leave behind CRs when deleting a Revision (or its parent package), I'm not sure we can assume that you want us to remove arbitrary finalizers that the provider controller may have added to the CRs, and I definitely don't think we should ever explicitly delete CRs. Do you have any thoughts here?

hasheddan replied:
I really like this idea. I'll get it written up in an issue for tracking :+1:

hasheddan replied:
Opened https://github.com/crossplane/crossplane/issues/1732 to track

negz commented on pkg/controller/pkg/revision/reconciler.go:
```diff
@@ -0,0 +1,565 @@
+	// or update any objects unless we can gain control of all objects 3. We
+	// will only modify owner references when the package revision is inactive
+	// 4. If objects are reconciled by a packaged controller, we will not remove
+	// owner references until the controller has been stopped
```
Nit: Looks like this list might have been wrapped.

negz reviewed (commented):
I didn't manage to make it all the way through this PR tonight, so I'll publish a partial review for now. It's looking good so far, though I'd like it if it was possible to lower the length and complexity of the reconcile loops by breaking some of the logic out into composable pieces.

hasheddan commented on apis/pkg/v1alpha1/configuration_types.go:
```diff
@@ -0,0 +1,114 @@
+	// Defaults to 1. Can be disabled by explicitly setting to 0.
+	RevisionHistoryLimit *int64 `json:"revisionHistoryLimit,omitempty"`
+
+	// ImagePullSecrets are named secrets in the same workspace that can be used
+	// to fetch Configurations from private repositories.
+	ImagePullSecrets []corev1.LocalObjectReference `json:"imagePullSecrets,omitempty"`
+
+	// ImagePullPolicy defines the pull policy for all images used during
+	// Configuration extraction.
+	// https://kubernetes.io/docs/concepts/configuration/overview/#container-images
+	ImagePullPolicy corev1.PullPolicy `json:"imagePullPolicy,omitempty"`
```
That sounds fine to me. It makes it a bit more abstract for the user, but since we are already intentionally introducing abstraction, I would be fine to continue that trend with these fields.

hasheddan commented on apis/pkg/v1alpha1/revision_types.go:
```diff
@@ -0,0 +1,124 @@
+	// package when the desired CRD is known, but the package name that contains
+	// it is not known. Either Package, CustomResourceDefinition, or Crossplane
+	// can be specified.
+	CustomResourceDefinition string `json:"crd,omitempty"`
```
I would be in favor of omitting until this functionality is actually available 👍

hasheddan commented on apis/pkg/v1alpha1/revision_types.go:
```diff
@@ -0,0 +1,124 @@
+
+	// Permissions are additional permissions that should be added to a packaged
+	// controller's service account.
+	Permissions *PermissionsSpec `json:"permissionSpec,omitempty"`
```
This is copied over from existing package types, and I think you are right that they should be `permissionRequests`. In practice, the RBAC controller would look at these and add them to the role it creates for the SA. Since they are part of the `Spec`, I would expect them to be populated by the `Package` controller, rather than the `PackageRevision` controller, but that involves parsing the package contents in the `Package` controller. How would you feel about this being moved to the status of `___Revision` like we did with Objects?

negz replied:
Regardless of what controller sets the requests, they do feel a little more like the desired state of the system to me. I can see why we might put them in status but I lean slightly toward keeping them in spec.

hasheddan replied:
The issue with putting them in `spec` is that if both the `Package` and `PackageRevision` controllers are watching the `PackageRevision`, then the `PackageRevision` controller updating the `spec` will trigger the the `Package` controller, which will then trigger the `PackageRevision` controller.... and so on. We are currently using the `APIPatchingApplicator`, and we could just say "if certain fields match, don't do a patch", but I would prefer if we were able to make it such that the `PackageRevision` controller didn't touch the `spec` at all. I do see your point that it feels like these belong in the `spec`, but it kind of depends on how you look at the revision. If the `spec` is indicating how it should parse and install a package, then you could argue that these don't belong in the `spec` since they are the _result_ of parsing the package, rather than indicating _how_ to parse a package.

negz replied:
> it kind of depends on how you look at the revision.

Yeah, on reflection I think your approach of putting these in status makes sense. If I think of "`spec.installPodRef` as being the desired state - or a pointer to the actual source of desired state (the install pod's logs) - then modelling all these as actual state in status makes sense.

hasheddan commented on pkg/controller/packages/hosted/sync.go:
```diff
@@ -59,10 +60,9 @@ func SyncImagePullSecrets(ctx context.Context, tenantKube, hostKube client.Clien
+	ref := &v1alpha1.TypedReference{
 		APIVersion: v,
 		Kind:       k,
-		Namespace:  hostNS,
```
I am going to move this out into a separate PR to reduce noise here 👍

hasheddan replied:
This has been factored out in https://github.com/crossplane/crossplane/pull/1718

hasheddan commented on pkg/controller/pkg/manager/reconciler.go:
```diff
@@ -0,0 +1,428 @@
+)
+
+// variableWait selects a shorter wait time if passed condition is true.
+func variableWait(short bool) time.Duration {
```
SGTM, I was attempting to achieve this with naming the parameter `short`, but I agree the name of the function is better practice. I wonder if a general function that took a boolean expression and two `time.Duration` would be useful across controllers 🤔

hasheddan commented on pkg/controller/pkg/manager/reconciler.go:
```diff
@@ -0,0 +1,428 @@
+	// is that the job will restart a failed pod automatically. However, if the
+	// job completes and the pod is then deleted, the job will not automatically
+	// recreate it. We handle this case below.
+	j := &batchv1.Job{}
```
I like the idea of just using a pod (hence this comment), but we do lose the functionality described in this comment.

negz replied:
Yeah. I didn’t think very hard about it, but I’m wondering if it would be simpler for us overall to just handle that functionality ourselves.

hasheddan commented on pkg/controller/pkg/manager/reconciler.go:
```diff
@@ -0,0 +1,428 @@
+			Applicator: resource.NewAPIPatchingApplicator(mgr.GetClient()),
+		},
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
```
Definitely, meant to include this in `TODO` along with more descriptive conditions. I figured these were somewhat trivial (relative to things like the API of the types and the controller design), so I deferred them to after initial review.

hasheddan commented on pkg/controller/pkg/manager/reconciler.go:
```diff
@@ -0,0 +1,428 @@
+			if err := r.client.Create(ctx, j); err != nil {
+				log.Debug(errCreateParserJob, "error", err)
+			}
+			p.SetConditions(runtimev1alpha1.Unavailable(), runtimev1alpha1.ReconcileSuccess())
```
Likely yes, but I think custom conditions may be in order here.

hasheddan commented on pkg/controller/pkg/manager/reconciler.go:
```diff
@@ -0,0 +1,428 @@
+
+	// If length of pod list is not 1, we are likely waiting for job to create pod.
+	if len(podList.Items) != 1 {
+		log.Debug(errGetParserPod, "error", errors.New("wrong number of pods for job"))
```
It is not, so would likely be good to keep this as a debug message but not as an error

hasheddan commented on pkg/controller/pkg/manager/reconciler.go:
```diff
@@ -0,0 +1,428 @@
+
+	// Do not create a revision if the image ID has not yet been populated.
+	if hash == "" {
+		log.Debug(errGetParserContainer, "error", errors.New("invalid image hash for container in parser pod"))
```
Same as comment above 👍

hasheddan commented on pkg/controller/pkg/revision/deployment.go:
```diff
@@ -0,0 +1,73 @@
+		},
+		Spec: appsv1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{"revision.pkg.crossplane.io": "true"},
```
SGTM

hasheddan commented on pkg/controller/pkg/manager/reconciler.go:
```diff
@@ -0,0 +1,428 @@
+	oldestRevisionIndex := -1
+	revisions := prs.GetRevisions()
+
+	// Check to see if revision already exists.
```
I'll look at doing so. However, I was a bit hesitant as the existing package manager is broken out into lots of smaller pieces and it personally feels hard to follow for me. I could see potentially breaking out some of the garbage collection and job / pod creation. I could also see having separate validation and create / update functions in the revision controller, similar to how we pass Applicators to our reconcilers. Let me refactor a bit and let you take another look 👍

negz replied:
That sounds like a good approach. It’s definitely tough to strike the balance between making the logic composeable and keeping it east to follow.

hasheddan commented on pkg/controller/pkg/manager/job.go:
```diff
@@ -0,0 +1,74 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
```
We have to take their address, so they cannot be `const`

hasheddan commented on pkg/controller/pkg/revision/deployment.go:
```diff
@@ -0,0 +1,68 @@
+		},
+		Spec: appsv1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{"pkg.crossplane.io/revision": revision.GetName()},
```
Very annoyingly, the limit on a k8s label value is 63 chars, while a sha256 is 64 characters :roll_eyes:

hasheddan commented:
@negz I am going to attempt to summarize the updates I have made since your last review to ease the burden of you next review pass:
- Switched to using `Pod` directly instead of `Job` for unpacking. The package controller acts sort of like an opinionated `Job` in that it will delete and recreate a `Pod` any time it enters an unhealthy state. I was trying to think of if we should always try to restart or if in some cases we should just say we tried and failed. I landed on always trying to restart but with a not very aggressive requeue. That way we shouldn't thrash too hard, but I suspect someone will either take action to update the package to a different version or take manual steps in the cluster to fix the install issue, so we might as well keep trying until we are told not to.
- Removed all fields for configuring a provider's controller for now from the `PackageRevisionSpec`
- Moved all fields that come from actually parsing a package to the status of the package revision to establish clear separation of responsibility between package controller and package revision controller (this include permissions and dependencies)
- Added custom conditions. I kind of struggled with what to use for conditions, but they are currently as follows:
  - `Installed`: this only applies to package types (i.e. `Provider` / `Configuration`). It is true if there the current `PackageRevision` is `Active`
  - `Healthy`: this applies to both package and package revision types. It is true for a package revision if all steps of installation for a package revision complete successfully. It is true for a package if the current package revision is healthy.
- Added events for errors and successful installation

There are still some unresolved comments above that I think we need to discuss a bit more. Outside of that, the following things remain:
- Finishing up unit tests (mostly waiting on this until the implementation is signed off on to avoid churn)
- Any factoring of logic into smaller functions. This will likely go hand-in-hand with finishing unit tests. It also could be partially addressed then followed up in a subsequent PR since it doesn't directly affect functionality.
- Rebase and clean up commits

Let me know if there are any other steps to make this PR a bit more manageable to review :)

hasheddan commented on apis/pkg/v1alpha1/provider_types.go:
```diff
@@ -0,0 +1,116 @@
+
+// ProviderSpec specifies details about a request to install a provider to
+// Crossplane.
+type ProviderSpec struct {
```
With `ControllerOptions` removed for now, I think we can move `ProviderSpec` and `ConfigurationSpec` to `PackageSpec` (similar to how we do with `PackageRevisionSpec`) and that can be the base that any future package type has to support (at minimum).

hasheddan commented on pkg/controller/pkg/manager/reconciler.go:
```diff
@@ -0,0 +1,401 @@
+		if kerrors.IsNotFound(err) {
+			pod = buildInstallPod(podName, r.namespace, p.GetSource(), map[string]string{parentLabel: p.GetName()}, p.GetPackagePullSecrets(), p.GetPackagePullPolicy())
+			meta.AddOwnerReference(pod, meta.AsOwner(meta.TypedReferenceTo(p, p.GetObjectKind().GroupVersionKind())))
+			if err := r.client.Create(ctx, pod); err != nil {
```
@kasey brought up a good question this morning about gc of completed Pods, which I wanted to add some context around...

`Succeeded` pods will be garbage collected according to the [`terminated-pod-gc-threshold` flag](https://kubernetes.cn/docs/concepts/workloads/pods/pod-lifecycle/#pod-garbage-collection) on the `kube-controller-manager`, which has a [default value of 12500](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podgc/config/v1alpha1/defaults.go#L34). This means that (with this current implementation), if an install Pod gets garbage collected (or manually deleted for that matter), the package revision will report that it is unhealthy, which will trigger a reconcile from the package controller, which will see that the Pod does not exist, and will recreate it. The package revision controller will then be successful on its next read of the Pod logs, will update to healthy, and that will be updated in the package as well.

This is assuming the current package revision though. If an _old_ package revision pod gets garbage collected, that revision will become unhealthy. This isn't a _huge_ problem, but it is a problem because an old current revision does still check that it has owner references on all of its installed objects, so it will fail to do that and the package controller will not recreate its Pod.

One way to get around this is to read the list of `Objects` we have on the revision (that get populated after reading Pod logs and modifying the objects), and if the package revision is inactive we can just use those to set the owner references (we are just adding them to the current object in the API server, so we don't need to know what it looks like in the package).

That being said, besides the old package revision being unhealthy, the fact that it is unhealthy is likely to not have effect on the system because nothing should be removing the owner references it has on an object, so it won't be updating them, but the object should still be garbage collected when the revision is deleted.

But don't worry, there's more complexity! If someone was to do a backup & restore of their cluster using something like Velero, the UID of the old package revision could change, and if it was the only owner reference on an object, that object would be GC'ed because it would see that none of its owners exist, and the Pod would not be created for the old package revision so it would not be able to go and update the owner references.

> aside: if there is no finalizer on the objects created by the package revision controller, could they be garbage collected on a backup and restore in the time it takes for the owner / controller references to get updated, even for the current package revision?

One thing that this all reaffirms in my mind is that storing objects in pod logs after initial unpack is a less than ideal situation. I hope that we can gain traction in a future iteration to unpack packages and either store them in volumes or some other medium.

hasheddan replied:
In regards to that last comment, another thing I would be in support of is getting the hash in the package controller (using something like `go-containerregistry`), but then deferring all unpacking and management of pods to the package revision controller.

negz commented on apis/pkg/v1alpha1/configuration_types.go:
```diff
@@ -0,0 +1,114 @@
+// configuration to Crossplane.
+type ConfigurationSpec struct {
+	// Package is the name of the package that is being requested, e.g.,
+	// myapp. Either Package or CustomResourceDefinition can be specified.
```
This was marked resolved, but no change was made. Was that intentional?

hasheddan replied:
Ah my apologies! Forgot to push those changes 👍

negz reviewed (commented):
Thanks @hasheddan! I've left some comments, but in general I think this PR is ready to be cleaned up for final approve and merge. I'd still like to make the reconcile logic a little more composable, but I think the priority at this stage should be adding tests and getting this merged for us to iterate on. Please make executive decisions as to whether you address my comments, choose to disregard them, or choose to defer them for future (though in the latter case please be sure to add a TODO or raise an issue).

While you wrap this PR up, would you mind opening a separate PR that adds the new package types? I think they're in pretty good shape, and getting them into master would make it a little easier to add provider support to the RBAC manager.

hasheddan commented on pkg/controller/pkg/manager/pod.go:
```diff
@@ -0,0 +1,68 @@
+						RunAsGroup:               &runAsGroup,
+						AllowPrivilegeEscalation: &allowPrivilegeEscalation,
+						Privileged:               &privileged,
+						RunAsNonRoot:             &runAsNonRoot,
```
We do not need to. A container will inherit from Pod if it is not set explicitly on the container. However, not all fields are available on the Pod, such as `Privileged` and `AllowPrivilegeEscalation`. If we wanted to strip some out here we should remove them from the Pod security context. I actually kind of like having them explicitly defined at both levels though because if someone is adjusting them in the future it feels like a stronger check if they have to change a value rather than add something net new (once again, we would have them defined at the container level and adding something to Pod security context would not override, so not super risky). However, this is mostly just a philosophical thing, so happy to simplify if desired.

hasheddan commented on pkg/controller/pkg/manager/reconciler.go:
```diff
@@ -0,0 +1,402 @@
+		// Find the oldest revision and delete it.
+		if err := r.client.Delete(ctx, gcRev); err != nil {
+			log.Debug(errGCPackageRevision, "error", err)
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, p), errUpdateStatus)
```
That's correct

hasheddan commented on pkg/controller/pkg/manager/reconciler.go:
```diff
@@ -0,0 +1,402 @@
+
+	// Set package to unavailable until package revision is available. Reconcile
+	// will be queued on changes to revisions, so we do not need to explicitly
+	// requeue here.
```
We previously were, although now we are not setting the `Healthy` status at all and just waiting to set it until the revision reports its health. Seems like having a blank `Healthy` status is better than showing `Unhealthy` before the revision reports back, but am open to suggestions here. Also, I'll update this comment.

hasheddan replied:
One caveat here: if it is not the first revision, then we will have the `Healthy` status of the old revision until the new one reports back.

hasheddan commented on pkg/controller/pkg/revision/reconciler.go:
```diff
@@ -0,0 +1,600 @@
+				exists:  false,
+			})
+
+			// If desired state is inactive, we attempt to dry run create the
+			// object with added owner reference to this package revision.
+			if desiredState == v1alpha1.PackageRevisionInactive {
+				meta.AddOwnerReference(d, meta.AsOwner(meta.TypedReferenceTo(pr, pr.GetObjectKind().GroupVersionKind())))
+				if err := r.client.Create(ctx, d, client.DryRunAll); err != nil {
+					log.Debug(errCreateDryRunObj, "error", err)
+					r.record.Event(pr, event.Warning(errCreateDryRunObj, err))
+					pr.SetConditions(v1alpha1.Unhealthy())
+					return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+				}
+				continue
+			}
+
+			// If desired state is active, we attempt to dry run create the
+			// object with added controller reference to this package revision.
+			if err := meta.AddControllerReference(d, meta.AsController(meta.TypedReferenceTo(pr, pr.GetObjectKind().GroupVersionKind()))); err != nil {
+				// NOTE(hasheddan): we would only hit this error if the object
+				// was packaged with a controller reference.
+				r.record.Event(pr, event.Warning(errEstablishControl, err))
+				pr.SetConditions(v1alpha1.Unhealthy())
+				return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+			}
+			if err := r.client.Create(ctx, d, client.DryRunAll); err != nil {
+				log.Debug(errCreateDryRunObj, "error", err)
+				r.record.Event(pr, event.Warning(errCreateDryRunObj, err))
+				pr.SetConditions(v1alpha1.Unhealthy())
+				return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+			}
+			continue
```
Yep! Although I assume we would want to check the error on the `meta.AsController`

hasheddan replied:
Ah nevermind, I see that is only for `AddControllerReference`. This works well for `Create` :+1:

hasheddan commented on pkg/controller/pkg/revision/reconciler.go:
```diff
@@ -0,0 +1,600 @@
+		return nil, nil
+	}
+	return nil, errors.New("unrecognized package meta type")
+}
```
This doesn't check that there aren't multiple package meta defined, which I think we should do. However, I am adding in linting so this function will go away anyway.

hasheddan commented on pkg/controller/pkg/revision/reconciler.go:
```diff
@@ -0,0 +1,600 @@
+		return nil, err
+	}
+	return metaScheme, nil
+}
```
I think I would prefer to bubble them up to the entrypoint, which would eventually call `FatalIfError` I believe.

hasheddan commented on pkg/controller/pkg/manager/pod.go:
```diff
@@ -0,0 +1,68 @@
+
+var (
+	runAsUser                = int64(2000)
+	runAsGroup               = int64(2000)
```
Only that they are typically not privileged. This can be helpful in the case that someone does not set their UID / GID in their container image, so it defaults to root when they don't really need it

hasheddan commented on pkg/controller/pkg/manager/reconciler.go:
```diff
@@ -0,0 +1,402 @@
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetPackage)
+	}
+
+	// Get existing package revisions.
```
Yep, I believe we can just defer to the GC here.

hasheddan commented on pkg/controller/pkg/manager/reconciler.go:
```diff
@@ -0,0 +1,402 @@
+	prs := r.newPackageRevisionList()
+	if err := r.client.List(ctx, prs, client.MatchingLabels(map[string]string{parentLabel: p.GetName()})); err != nil {
+		log.Debug(errListRevisions, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errListRevisions)
```
I believe `kmeta.IsNoMatchError` is returned if the API types is not able to be recognized by the client. The doc on `List` says:
```
// List retrieves list of objects for a given namespace and list options. On a
// successful call, Items field in the list will be populated with the
// result returned from the server.
``` 
Which sound to me like it should return an error if none are found. However, I actually don't know how what I have right here would be working if that was the case because we would never get past this point because we would start with no package revisions, and then we would never create any :thinking:

hasheddan commented on pkg/controller/pkg/revision/linter.go:
```diff
@@ -0,0 +1,166 @@
+	errNotComposition       = "object is not a Composition"
+)
+
+// A Linter lints packages.
```
I anticipate moving much of this (excluding the functions specific to Crossplane types) to `crossplane-runtime` at some point

negz commented on pkg/controller/pkg/revision/establisher.go:
```diff
@@ -0,0 +1,178 @@
+		c, ok := current.(resource.Object)
+		if !ok {
+			return errors.New(errAssertObj)
+		}
```
I'm pretty sure this branch should never be hit, since we check at line 83.

negz commented on pkg/controller/pkg/revision/establisher.go:
```diff
@@ -0,0 +1,178 @@
+	return nil
+}
+
+// Establish establishes control or ownership of resources in the package.
```
Could be worth clarifying that the caller must call Check before calling Establish.

negz commented on pkg/controller/pkg/revision/reconciler.go:
```diff
@@ -0,0 +1,398 @@
+	namespace    string
+	client       resource.ClientApplicator
+	establisher  Establisher
+	podLogClient *kubernetes.Clientset
```
Nit: Could this be a `kubernetes.Interface`?

negz commented on pkg/controller/pkg/revision/reconciler.go:
```diff
@@ -0,0 +1,398 @@
+	// true.
+	pr.SetObjects(r.establisher.GetResourceRefs())
+	// TODO(hasheddan): set remaining status fields (dependencies, crossplane
+	// version, permission requests)
```
Does this TODO need addressing before we merge, or do you want to handle it in a future PR?

hasheddan replied:
Future PR :+1: Related to how RBAC controller will create roles for controllers

negz reviewed (approved):
🎉

hasheddan commented on pkg/controller/pkg/manager/reconciler_test.go:
```diff
@@ -0,0 +1,713 @@
+							return nil
+						}),
+					},
+					podManager: NewPackagePodManager(&test.MockClient{
```
Agreed :+1:

hasheddan replied:
Updated both here and for the package revision establisher :+1:
